gpu matrix multiply

SteveBronder opened the pull request:
## Summary

Adds functions for matrix multiplication on the GPU for:
1. `matix_gpu * matrix_gpu`
2. `matrix_gpu * scalar`
3. `matrix_gpu * matrix_gpu^T`

## Tests

1. `diagonal_multiply_test`
2. `multiply_transpose_test`
3. `multiply_test`

These tests all check for basic compilation and correctness either with known values or against the CPU version. Where domain_errors are caught we add an appropriate test, however errors from `check_opencl_error()` are system_errors caused by hardware or the OpenCL platform failing and so are not caught in the test cases.

Reviewers can run the tests with

```bash
echo STAN_OPENCL=true>> make/local
echo OPENCL_PLATFORM_ID=0>> make/local
echo OPENCL_DEVICE_ID=0>> make/local
make test/unit/math/gpu/diagonal_multiply_test && ./test/unit/math/gpu/diagonal_multiply_test
make test/unit/math/gpu/multiply_transpose_test && ./test/unit/math/gpu/multiply_transpose_test
make test/unit/math/gpu/multiply_test && ./test/unit/math/gpu/multiply_test
```

## Side Effects

None

## Checklist

- [x] Math issue #974 

- [x] Copyright holder: 

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
Rok Češnovar and Erik Štrumbelj (Faculty of Computer and Information Science, University of Ljubljana)
Steve Bronder

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

SteveBronder commented:
It seems like we are failing the performance tests

http://d1m1s1b1.stat.columbia.edu:8080/blue/organizations/jenkins/Stan/detail/downstream_tests/73/pipeline

Is this a known upstream issue or are we the cause of this? That test doesn't have the flag `STAN_OPENCL` so I'm a bit confused with how to diagnose what's wrong if its because of the GPU code

SteveBronder commented:
I'll look into this tonight

SteveBronder commented:
@syclik @seantalts tests are passing so this is ready for review!

seantalts commented:
Should this wait for #966? Seems like they'd have conflicts between them

rok-cesnovar commented:
Yes, the way that the kernels are retrieved will change once #966 is merged. 

`cl::Kernel kernel = opencl_context.get_kernel("scalar_mul");`
will be changed to 
`kernel_cl kernel("add");`

Other than that, there will be no changes.

Should we close and then reopen it?

seantalts commented:
I think it's up to you how to stage it - I'd say serializing them (doing one and then the other) is preferred because of the way the two PRs are operating over some of the same code, your choice of which one to put first.

rok-cesnovar commented:
Will reopen once #966 is merged.

seantalts commented:
Should we resurrect this? Who wants to do the honors?

rok-cesnovar commented:
I will merge the kernel changes and then reopen.

rok-cesnovar commented:
I merged the recent kernel changes. 

I added the `local_range_kernel` functor that expects 2 cl::NDRange arguments for the thread block size.

What would be the best way to handle the parameters that are set [here](https://github.com/stan-dev/math/pull/967/files#diff-0e228daae4e50272f0d16ec700ee8c34R33)?

Foro the `WORKGROUP_SIZE_MULT` we need to put a check finds `min(sqrt(max_workgroup_size()), 32)` and then sets the the `WORK_PER_THREAD` to 1 if `WORKGROUP_SIZE_MULT` is not 32.

Should we do this in `compile_kernel()`?

rok-cesnovar commented on stan/math/gpu/kernel_cl.hpp:
```diff
@@ -29,24 +29,46 @@ map_base_opts base_opts
+  // dimension of a square workgroup size
+  // WG size of (32,32) works on all recent GPU but would fail on some
+  // older integrated GPUs or CPUs
+  if (wg_size_sqrt < base_opts["WG_SIZE_MULT"]){
```
Do we want to do this here? I dont have any better ideas right now.

seantalts replied:
Hmmm. Can we make this stuff live on the opencl_context and be computed just once when the context is initialized? It can then expose the options map as part of it's API perhaps?

rok-cesnovar replied:
Tnx! Seems like a good idea.

rok-cesnovar commented on stan/math/gpu/kernels/helpers.hpp:
```diff
@@ -32,6 +32,23 @@ std::string helpers =
   #ifndef dst
   #define dst(i,j) dst[j * dst_rows + i]
   #endif
+  #ifndef WORK_PER_WI_MULT
```
Should we make a separate file for kernel constants?

seantalts replied:
Eh, I think it's fine like this until this file becomes super huge. It's nice that developers looking at / working on kernels have only one file of predefined stuff to look at.

rok-cesnovar commented on stan/math/gpu/multiply.hpp:
```diff
@@ -0,0 +1,92 @@
+  matrix_gpu temp(A.rows(), B.cols());
+  if (temp.size() == 0)
+    return temp;
+  int local = base_opts["WG_SIZE_MULT"];
```
I dont like this. We need to be able to check this at compile time.

rok-cesnovar replied:
`base_opts` is defined inside a anonymous namespace which makes it local inside a translation unit. But based on your other comment this will get moved to the opencl_context anyways.
```
namespace stan {
namespace math {
namespace {
```

seantalts replied:
Woops, I didn't realize that was global to the entire translation unit - I thought just the file! Eep. While you're in there would you also mind making that one `internal` instead and maybe naming the stuff better? Or I can do that in this or another PR if it's easier.

rok-cesnovar replied:
What if I moved it to the `stan::math::opencl_kernels` namespace? Since this will only contain the compile_kernel function (the base_opts will be moved to context).

I would want to avoid 
```
public:
  kernel_functor(const char* name, const char* source)
      : kernel_(stan::math::internal::compile_kernel(name, source)) {}
```

rok-cesnovar replied:
I am not sure at which stuff you are aiming at that you say should be named better? The constants in `base_opts or the variables inside this matrix_multiply?

seantalts replied:
the opencl_kernels namespace works for me for this stuff. Re: names - I just meant the `compile` function and `base_opts` etc that were exposed to the entire translation unit (all of Stan Math is one translation unit typically, so that was just totally global, haha). My bad.

rok-cesnovar commented on stan/math/gpu/kernel_cl.hpp:
```diff
@@ -29,24 +29,46 @@ map_base_opts base_opts
   } catch (const cl::Error& e) {
+    std::string buildlog = program.getBuildInfo<CL_PROGRAM_BUILD_LOG>(
+        opencl_context.device()[0]);
+    std::cerr << "Build log :" << std::endl << buildlog << std::endl;
```
Writing to std::cerr is probably not OK. Where can I put this? I would really like to output the logs in case of a kernel compile error. Just giving out error -11 is not really informative.

seantalts replied:
I think we normally just throw exceptions as our primary way of communication on error conditions. Can you wrap up the build log into an exception to throw?

rok-cesnovar commented:
I made a few comments. 

One general thing we have to discuss is whether we are going to use thread/thread block or workitem/workgroup. 

The former is much more informative to me and I think a clearer term but the latter is OpenCLs way of describing it. I would vote thread/thread block.

SteveBronder commented:
Sorry I've been sort of lazy all day and saw Rok making things so I thought I would wait. Going to look this over right now.

I prefer thread and threadblock.

SteveBronder commented:
Did a couple little things

1. Since the `THREAD_BLOCK_SIZE_FUNC` was the same for both of the multiplications I removed them and made one macro `THREAD_BLOCK_SIZE`
2. Changed the macro names from referencing work groups to reference threads and thread groups
3. Added documentation to the kernel functor stuff (we forgot that last time)
4. Added the LOWER, UPPER, ENTIRE, etc. macros into the `constants.hpp` file then removed them from the compilation macro map. At this point just the thread block and work per thread macros remain in that map. We don't have to tackle it in this PR but we are pretty close to getting rid of that map.

@rok-cesnovar how did you decide the choice of work per thread? To remove that macro we could add an argument to the kernel compilation. Then we would over-ride that value if the max thread block size is too small (that sounds weird to me though)

I have not read through the kernels yet, but everything else looks good to me!

SteveBronder commented on stan/math/gpu/kernels/matrix_multiply.hpp:
```diff
@@ -0,0 +1,88 @@
+                                  const int M, const int N, const int K) {
+      // workitem index inside the workgroup
+      const int workgroup_row = get_local_id(0);
+      const int workgroup_col = get_local_id(1);
```
Should these all be changed to size_t? and would a better name here be workitem_row/col?

rok-cesnovar replied:
Maybe something like local_row, local_col would be better. These are indices inside a workgroup/thread block, thats why I put that prefix there.

rok-cesnovar replied:
Agree on the size_t comment. I will change this together with the changes to thread/thread block naming conventions as I will have to go through all the kernels anyway.

SteveBronder commented on stan/math/gpu/kernels/matrix_multiply.hpp:
```diff
@@ -0,0 +1,89 @@
+        const read_only int N, const read_only int K) {
+      // workitem index inside the workgroup
+      const int workgroup_row = get_local_id(0);
+      const int workgroup_col = get_local_id(1);
```
Should these the the workitem rows/cols?

Also should we change all the ints, unsigned ints, etc. that we use in the kernels for getting indices into `size_t`?

rok-cesnovar replied:
See previous comment.

SteveBronder commented on stan/math/gpu/kernels/matrix_multiply.hpp:
```diff
@@ -0,0 +1,89 @@
+          const int tiled_j = THREAD_BLOCK_SIZE * tile_ind + workgroup_col;
+
+          A_local[workgroup_col + w * THREAD_BLOCK_SIZE_MULT_COL][workgroup_row]
+              = A[(tiled_j + w * THREAD_BLOCK_SIZE_MULT_COL) * M + i];
```
I think we could use the macro for the matrices here if we change them to something like
```cpp
#define A(i, j, k) A[j * k + i]
```

Though I think I'd like to wait till another PR to do that

rok-cesnovar commented:
Great! Thanks!

> Added the LOWER, UPPER, ENTIRE, etc. macros into the constants.hpp file then removed them from the compilation macro map. At this point just the thread block and work per thread macros remain in that map. We don't have to tackle it in this PR but we are pretty close to getting rid of that map.

I am not sure we can do this. The map was added so that the LOWER/UPPER/ENTIRE constants are all defined in one place. The example that Daniel gave that time was that if someone decides to change a constant [here](https://github.com/bstatcomp/math/blob/gpu_matrix_multiply/stan/math/gpu/constants.hpp#L6) he doesnt have to go through all the occurences of LOWER/UPPER/ENTIRE in the kernels to have this working.

rok-cesnovar commented:
> @rok-cesnovar how did you decide the choice of work per thread? To remove that macro we could add an argument to the kernel compilation. Then we would over-ride that value if the max thread block size is too small (that sounds weird to me though)

Me and Davor did a somewhat thorough analysis of this. We ran OpenCL benchmarks for the matrix multiplication kernels without data copy on NVIDIAs Tesla V100, Tesla K20m, Titan XP, GTX 760 as well as AMDs HD5750 and R9 290x. On all newer/high end GPUs the 32/8 combination was the fastest and on the  GT760 and HD5750 it was the second fastest by 2-3% of kernel execution time, which means even less if we add the data copy.

We also tried some other optimizations like using vectored data types and register tiling but they were all too architecture dependent. The fastest implementation on certain GPUs with those optimizations were slower than without optimizations on others.

Run-time optimizations and auto-tuning GPU implementations is something I want to try here someday because Stan model executions are perfect cases for that, but we are nowhere near that time haha :) Things like Stan helping Stan get faster sounds pretty fun to me :)

> Then we would over-ride that value if the max thread block size is too small (that sounds weird to me though)
Imagine a case where a max 8 x 8 thread block is allowed, in that case we would spawn 8x1 threads which is not optimal. It could also be possible that a max 4x4 thread block is allowed, in which case work-per-thread of 8 would not be allowed. We could further complicate this check and cover all the middle ground like a `16x16` thread-block size would have `work_per_thread = 4`. I do think that we just need to make sure that the GPU code doesnt break in these cases. The kernel will be 10-20% slower if we dont tune for them but those older GPUs will be slower anyways.

seantalts commented:
> I am not sure we can do this. The map was added so that the LOWER/UPPER/ENTIRE constants are all defined in one place. The example that Daniel gave that time was that if someone decides to change a constant here he doesnt have to go through all the occurences of LOWER/UPPER/ENTIRE in the kernels to have this working.

Hmm, this is a good point, though now that we have this shared `helpers.hpp` at least it is just one place instead of in each kernel. I'm a little bit on the side of keeping them in the map here actually, mostly because I'm now thinking of a map like this as necessary because it contains not just compile time constants like `UPPER` but variables that could change once the code is running on a specific computer with specific hardware - variables like thread group size and work per thread that might even be tuned as a Stan program is running.

SteveBronder commented:
> I'm a little bit on the side of keeping them in the map here actually, 
That makes sense I can add them back. @rok-cesnovar do we also want to keep the THREAD_BLOCK per kernel or are you okay with having one?

rok-cesnovar commented:
I would maybe rather have two as these two kernels differ so much that we could potentially have different parameter value.

SteveBronder commented:
How about instead of having a thread block size per kernel we change the constructor for local and global range kernels so we can pass in a map to override / add to the macros before we compile.

```cpp
const local_range_kernel<cl::Buffer, cl::Buffer, int, int> multiply_transpose(
    "multiply_transpose", multiply_transpose_kernel_code, 
 {{'THREAD_BLOCK_SIZE', 12}, {'WORK_PER_THREAD', 4});
```

I think we would need something like this to tune the kernels in the future anyway.

SteveBronder commented:
Excellent! I'll tackle this after work

SteveBronder commented:
Odd, the branch is showing my latest update but it's not on this PR? (just updated to be in line with master)

https://github.com/bstatcomp/math/tree/gpu_matrix_multiply

SteveBronder commented:
Seeing 
```
[Stan_downstream_tests-6WZJY57OMM7SR23VI3JNDCM2GFWGM7AM6DG3JZQFFBGQGVZ7OSRA] Running batch script
C:\Jenkins2\workspace\Stan_downstream_tests-6WZJY57OMM7SR23VI3JNDCM2GFWGM7AM6DG3JZQFFBGQGVZ7OSRA>make -j16 test-headers 

'make' is not recognized as an internal or external command, operable program or batch file.
```
I'm guessing the windows jenkins comp needs to find the cygwin make. We can rerun our stuff once that gets updated

seantalts commented:
The windows box was broken, I just restarted the upstream tests portion, hopefully that works.

seantalts commented on stan/math/gpu/kernels/scalar_mul.hpp:
```diff
@@ -0,0 +1,45 @@
+                             const unsigned int cols) {
+      int i = get_global_id(0);
+      int j = get_global_id(1);
+      if (i < rows && j < cols) {
```
Same question here - can we not rely on the NDRange to do the bounds checking?

SteveBronder replied:
@rok-cesnovar correct if if I'm wrong here, but not we can't. If we have a work group of size (32,32) and global of (34,34) that means 4 work groups would go off (the main chunk to handle index (0:31) then 3 others to handle indexes for (31:32). So we need this if statement to make sure those 3 work groups only work within the matrix

rok-cesnovar replied:
You are right @SteveBronder. In these cases we dont even know what is the thread block size, as we leave it to the OpenCL implementation to handle the thread-block size. 

In order to avoid this we would have to set the thread-block size everywhere and pad the matrices the way we do it in the matrix multiplication. Since these kernels are pretty simple this checks do not present any real performance issues.

seantalts commented on test/unit/math/gpu/diagonal_multiply_test.cpp:
```diff
@@ -0,0 +1,66 @@
+#ifdef STAN_OPENCL
+TEST(MathMatrix, multiply_m_exception_pass_diagonal_mul) {
+  stan::math::matrix_d m0;
+  m0.resize(3, 2);
```
I think you can accomplish this with
```cpp
stan::math::matrix_d m0(3, 2);
```

seantalts commented on test/unit/math/gpu/diagonal_multiply_test.cpp:
```diff
@@ -0,0 +1,66 @@
+TEST(MathMatrix, multiply_m_exception_pass_diagonal_mul) {
+  stan::math::matrix_d m0;
+  m0.resize(3, 2);
+  m0 << 1, 1, 1, 1, 1, 1;
```
The last 3 lines together are 
```cpp
auto m0 = MatrixXd::Ones(3, 2);
```

seantalts commented on test/unit/math/gpu/diagonal_multiply_test.cpp:
```diff
@@ -0,0 +1,66 @@
+  stan::math::matrix_d m3_dst;
+  m3.resize(3, 3);
+  m3_dst.resize(3, 3);
+  m3 << 2, 2, 2, 2, 2, 2, 2, 2, 2;
```
```cpp
auto m3 = stan::math::matrix_d::Ones(3, 3) * 2;
```

seantalts commented on test/unit/math/gpu/diagonal_multiply_test.cpp:
```diff
@@ -0,0 +1,66 @@
+  stan::math::matrix_gpu mm3_dst(mm3.rows(), mm3.cols());
+  EXPECT_NO_THROW(mm3_dst = stan::math::diagonal_multiply(mm3, 3.0));
+  stan::math::copy(m3_dst, mm3_dst);
+  EXPECT_EQ(6.0, m3_dst(0, 0));
```
I would really like to see these tests compare against the standard stan math functions that do the same thing. This might require a few utility helper functions to e.g. EXPECT_FLOAT_EQ two Eigen Matrices.

rok-cesnovar replied:
That is a great idea. Lets add the helper in this PR as it will be useful in the upcoming PRs also.

seantalts replied:
I think generally the form for these GPU tests could look something like generating a bunch of random matrices and checking gpu_multiply vs. multiply, though we'd also want to hand write some edge-cases. There are fun toolkits that generate edge cases automatically but we don't have one yet in Stan...  like https://github.com/emil-e/rapidcheck

seantalts commented on test/unit/math/gpu/multiply_test.cpp:
```diff
@@ -0,0 +1,284 @@
+  stan::math::matrix_gpu rvv(rv);
+  stan::math::matrix_gpu vv(v);
+  stan::math::matrix_gpu ans_vv(1, 1);
+  EXPECT_NO_THROW(ans_vv = stan::math::multiply(rvv, vv));
```
Why are there so many EXPECT_NO_THROW tests? I typically think of these as something to pop in once you have seen bugs and are coding a regression test for specific bugs. Did someone recommend these?

rok-cesnovar replied:
Some of those were added just so that we checked whether anything breaks in edge cases.

seantalts commented on test/unit/math/gpu/multiply_test.cpp:
```diff
@@ -0,0 +1,284 @@
+
+TEST(MathMatrix, multiply_m_v_exception_pass) {
+  stan::math::matrix_d m;
+  stan::math::vector_d v;
```
These are never initialized and are probably pointing to random data, which could be NaN or something and possibly end up throwing an exception (at random), right? Can you initialize them with something?

rok-cesnovar replied:
We simply just copied most of these tests from the multiply_test. For example [here](https://github.com/stan-dev/math/blob/develop/test/unit/math/prim/mat/fun/multiply_test.cpp#L102)

seantalts commented on test/unit/math/gpu/multiply_test.cpp:
```diff
@@ -0,0 +1,284 @@
+  vector_d vec(4);
+  vec << 1, 2, 3, 4;
+  vector_d t_vec(4);
+  assign(t_vec, multiply(vec, d_int));
```
It looks like this test (and some others in here) don't even use the GPUs?

seantalts commented on test/unit/math/gpu/multiply_transpose_test.cpp:
```diff
@@ -0,0 +1,125 @@
+
+  for (int i = 0; i < size_x; i++)
+    for (int j = 0; j < size_x; j++)
+      EXPECT_NEAR(m2(i, j), m2_gpu(i, j), 1e-10);
```
This is good, we could maybe write a util for it

SteveBronder commented on stan/math/gpu/opencl_context.hpp:
```diff
@@ -271,13 +288,20 @@ class opencl_context {
    */
-  inline int max_workgroup_size() {
-    return opencl_context_base::getInstance().max_workgroup_size_;
+  inline int max_thread_block_size() {
```
idt it's a bad idea to let users have access to this and could see it being useful in the future. Though not necessary

SteveBronder commented on stan/math/gpu/kernels/matrix_multiply.hpp:
```diff
@@ -0,0 +1,92 @@
+      }
+      // save the values
+      for (int w = 0; w < WORK_PER_THREAD; w++) {
+        // each thread saves WORK_PER_THREAD_SELF_TRANS values
```
Ah, missed this when fixing the WORK_PER_THREAD names

rok-cesnovar replied:
Cleaned up.

SteveBronder commented on stan/math/gpu/kernels/multiply_transpose.hpp:
```diff
@@ -0,0 +1,108 @@
+            A_local[thread_block_col + w * THREAD_BLOCK_SIZE_COL]
+                   [thread_block_row]
+                = A[i + (tiled_j + w * THREAD_BLOCK_SIZE_COL) * M];
+            B_local[thread_block_col + w * THREAD_BLOCK_SIZE_COL]
```
As of now we compile each kernel separately so this kernel can't use another one. Though we could pass dependent kernels as an argument

rok-cesnovar commented on stan/math/gpu/kernels/scalar_mul_diagonal.hpp:
```diff
@@ -0,0 +1,43 @@
+        __global read_write double *A, const read_only double scalar,
+        const read_only unsigned int rows, const read_only unsigned int cols) {
+      int i = get_global_id(0);
+      if (i < rows && i < cols) {
```
We need the `rows` because A(i,i) is actually `A[ j * rows + i ]`.

rok-cesnovar replied:
I changed this so that it `rows` and `min_dim` are passed in. The check is therefore `(i < min_dim)`.

rok-cesnovar commented on stan/math/gpu/kernel_cl.hpp:
```diff
@@ -20,61 +19,139 @@
   }
   std::string kernel_source(opencl_kernels::helpers);
   kernel_source.append(source);
+  cl::Program program;
```
Because we need this in the catch section, I had to move it outside of the try section.

seantalts replied:
Ah ok, yeah. Oops.

rok-cesnovar commented on stan/math/gpu/opencl_context.hpp:
```diff
@@ -271,13 +288,20 @@ class opencl_context {
-   * work groups of sizes (16,16), (128,2), (8, 32), etc.
+   * Returns the reference to the map of kernel defines
+   */
+  inline opencl_context_base::map_base_opts& base_opts() {
```
fixed

rok-cesnovar commented on stan/math/gpu/kernels/matrix_multiply.hpp:
```diff
@@ -0,0 +1,92 @@
+                               [block_ind];
+          }
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
```
Yes, that helps, but more importantly, we do not want to overwrite any existing values in the local/shared memory until all the threads calculate read them and calculate their `acc`.

rok-cesnovar commented on stan/math/gpu/multiply.hpp:
```diff
@@ -0,0 +1,92 @@
+ * @throw <code>std::invalid_argument</code> if the
+ *   number of columns in A and rows in B do not match
+ */
+inline matrix_gpu multiply(matrix_gpu& A, matrix_gpu& B) {
```
I added const here and in the other multiply functions (where applicable).

rok-cesnovar commented on stan/math/gpu/kernel_cl.hpp:
```diff
@@ -20,61 +19,139 @@
-  global_range_kernel(const char* name, const char* source)
-      : make_functor(name, source) {}
-  auto operator()(cl::NDRange thread_block_size, Args... args) const {
+  std::map<const char*, int> opts;
```
Yep. Added const to all the cases below also.

rok-cesnovar commented:
I think I managed to address most of the comments. Things that are still open:

- refactoring the tests
- moving the merged opts as the third `compile_kernel`

Would these tests be ok?

Edge cases that should not throw
- vector x row vector
- multiply Nx0 and 0xM
- multiply 0xM and Nx0
- multiply 0x0 and 0x0
- multiply matrices Nx0, 0xN and 0x0 with a scalar; both `multiply(A, scalar)` and `multiply(scalar, A)`

Edge cases that should throw: 
- non matching dimensions tests

Tests with random matrices compared with the existing Stan Math multiply:
-`multiply(A, scalar)` and `multiply(scalar, A)`  for vector, row_vector and matrix
- row-vector x vector and row-vector x vector
- small matrices
- large matrices

seantalts commented:
I think those tests are fine. I just think going forward we might want a new pattern for this style of function that is supposed to be exactly the same as some other expression but with better performance. Would like someone (maybe me) to play around a little with this rapidcheck library and see if it makes sense to use that to check these properties.

seantalts commented:
This is ready again?

seantalts commented:
Word, thanks. I'm looking at the opts thing now and playing around with it...

seantalts commented:
What do you guys think of this? 

Will review the rest of it now!

seantalts commented on test/unit/math/gpu/diagonal_multiply_test.cpp:
```diff
@@ -5,12 +5,8 @@
-  stan::math::matrix_d m1;
-  m1.resize(1, 3);
-  m1 << 1, 1, 1;
+  stan::math::matrix_d m0 = stan::math::matrix_d::Ones(3, 2);
```
You can also do `auto m0 = stan::math::matrix_d::Ones(3, 2);` and you could do a `using stan::math::matrix_d;` at the top of a test file if you wanted. No need to change this now.

seantalts commented:
I think it's good to go, but would like one of you guys to say if you like this version of the opts thing. The major semantic changes I wanted to get in re: opts were:
1. capture the full list of ones used to compile the kernel (meaning merging in the base_opts at the time of compilation), and
2. make access to them immutable from outside code so no one accidentally tampers with them.

rok-cesnovar commented:
I really like this version and definitely have not better idea of doing this.

seantalts commented:
Okay, let's go with this for now. Merging!

wds15 commented:
whooo.... are you now shooting for the cholesky and then we release this?

seantalts commented:
I'm not sure - would ideally also have a mechanism to avoid copying data matrices to the GPU every leapfrog step (one idea for that outlined in https://github.com/stan-dev/math/pull/1014).

rok-cesnovar commented:
The next PR is going to be the lower triangular inverse (I will open it today or tomorrow) and then the Cholesky PR. The latter one will have very little new GPU code as it will use all the merged GPU functions.

The other thing we will also try to tackle in the meantime is reducing the data transfer Sean mentioned. The other thing that we need to figure out is the data transfer on input/ouput of the functions.

The way we have things prepared for the cholesky decompose now, it will accept Eigen matrices as input and will output an Eigen matrix. Everything in between happens on a GPU. If the cholesky decompose is the only GPU function this really isnt a problem, but if we look at

```
matrix[N, N] cov =   cov_exp_quad(x, alpha, rho)
                     + diag_matrix(rep_vector(square(sigma), N));
matrix[N, N] L_cov = cholesky_decompose_gpu(cov);

rho ~ inv_gamma(8.91924, 34.5805);
alpha ~ normal(0, 2);
sigma ~ normal(0, 1);

y ~ multi_normal_cholesky(rep_vector(0, N), L_cov);
```
The optimal way would be that `x`, `alpha`, `rho` and `sigma` are transferred to the GPU and `y` is transferred back to the CPU memory. This means N+3 doubles are transfered to the GPU and N doubles are transferred back. Of course this would only work if all the above functions have their GPU counterparts. If, for instance, the multi_normal_cholesky would not be done on the GPU, we would transfer back after the cholesky decompose (N^2 doubles to transfer back).
But this is still a lot less than if we transfer back and forth in all functions.

seantalts commented:
I think the test is failing: http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pipeline/job/develop/160/testReport/junit/(root)/MathMatrix/Additional_merge_tests___Unit_with_GPU___row_vector_vector/

SteveBronder commented:
Sorry for being mia, back now.

Is that test failing because of a difference in hardware?

seantalts commented:
Could be! I never ran the tests locally as I haven't gotten GPU stuff set up on my Windows machine yet and my normal laptop is in the shop. Trying to get the Windows box set up but it's failing to link opencl right now.

SteveBronder commented:
If there is anything I can do to help lmk! Going to clean up some of the lower triangular inverse code rn and make a wip pull request

seantalts commented:
If you're just looking for helpful stuff to do, I probably won't be able to
fix the tests soon so if you or rok can get to it that would be cool. But
also enjoy your weekend!
On Sat, Sep 8, 2018 at 18:09 Steve Bronder <notifications@github.com> wrote:

> If there is anything I can do to help lmk! Going to clean up some of the
> lower triangular inverse code rn and make a wip pull request
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/967#issuecomment-419676050>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7GXbQi-uWWd2UQ3ARkzALjcuy4Wtks5uZEACgaJpZM4VvgWN>
> .
>
-- 
📲

SteveBronder commented:
Mkay I can get this to replicate on a mac. I would guess this is because of the OpenCL driver that comes with mac but will look into it more

SteveBronder commented:
Actually I hope this is not failing because of the reasons it's failing on my mac. Can you post the output of `clinfo` as a gist?

Mine is failing because I only have a small intel graphics card that does not support double precision floating points. The other issue could be that we need to use the OpenCL headers that macs come with, though that's not bad.

```cpp
#if defined(__APPLE__) || defined(__MACOSX)
#include <OpenCL/cl.hpp>
#else
#include <CL/cl.hpp>
#endif
```

rok-cesnovar commented:
@SteveBronder hopefully you didnt start with the lower trianuglar inverse cleanup. I am already 3/4 through this. I was planning on having this done on Friday so I didnt sent out any email, but then started with a bigger refactor as I had a new idea for a more clean implementation. It should be done today, I will let you know.

We are including the OpenCL headers in Stan Math, so the #if defined (__APPLE__) tricks should not be needed as we do not look for the systems OpenCL headers.

If the device is OpenCL 1.0, you may need to manually turn on double precision support like this:
```
#ifdef cl_khr_fp64
#pragma OPENCL EXTENSION cl_khr_fp64 : enable
#elif defined(cl_amd_fp64)
#pragma OPENCL EXTENSION cl_amd_fp64 : enable
#else
#error "Double precision floating point not supported by OpenCL implementation."
#endif
```
We had this in the kernels at one point, but we removed it at some point. Should we just add this back to the helpers?

seantalts commented:
Here is the device info (there was no clinfo command):
```
1. Device: Intel(R) Xeon(R) CPU E5-1680 v2 @ 3.00GHz
 1.1 Hardware version: OpenCL 1.2 
 1.2 Software version: 1.1
 1.3 OpenCL C version: OpenCL C 1.2 
 1.4 Parallel compute units: 16
2. Device: AMD Radeon HD - FirePro D700 Compute Engine
 2.1 Hardware version: OpenCL 1.2 
 2.2 Software version: 1.2 (May  2 2018 22:40:03)
 2.3 OpenCL C version: OpenCL C 1.2 
 2.4 Parallel compute units: 32
3. Device: AMD Radeon HD - FirePro D700 Compute Engine
 3.1 Hardware version: OpenCL 1.2 
 3.2 Software version: 1.2 (May  2 2018 22:40:03)
 3.3 OpenCL C version: OpenCL C 1.2 
 3.4 Parallel compute units: 32
```

I tried switching from the Xeon to the GPU by changing the device id to 1 but the tests still fail...

seantalts commented:
@rok-cesnovar  I tried adding that bit above about manually enabling double precision and it didn't help. FWIW the matrices are coming back entirely `nan`, which seems like something more than a precision problem...?

seantalts commented:
I figured it out - when we're doing the padding, we're not zeroing out the padded stuff at all, so it could be 1) random data 2) helpfully ish, zeros (nvidia seems to zero it out for us), or 3) `nan` to alert us that we're using uninitialized memory (thanks AMD). I actually think that's kinda cool of them. Zeroing out the temp buffers before using them works... not sure if it can be made more efficient.

seantalts commented:
what exactly is a matrix with 0 rows or columns? Why shouldn't that throw an exception?

seantalts commented:
I think I have a fix in #1027 - let's continue there

SteveBronder commented:
> hopefully you didnt start with the lower trianuglar inverse cleanup

I decided to wait until the gpu_matrix_multiply is merged so all is well
